cumulus/test/service/src/chain_spec.rs

```diff
@@ -50,17 +50,39 @@ pub fn get_chain_spec_with_extra_endowed(
 	extra_endowed_accounts: Vec<AccountId>,
 	code: &[u8],
 ) -> ChainSpec {
+	let mut patch_json = json!({
+		"balances": {
+			"balances": extra_endowed_accounts.into_iter().map(|a| (a, 1u64 << 60)).collect::<Vec<_>>(),
+		}
+	});
+
+	if let Some(id) = id {
+		// Merge parachain ID if given, otherwise use the one from the preset.
+		sc_chain_spec::json_merge(
+			&mut patch_json,
+			json!({
+				"parachainInfo": {
+					"parachainId": id,
+				},
+			}),
+		);
+	};
+
+	let runtime_caller = GenesisConfigBuilderRuntimeCaller::<ParachainHostFunctions>::new(code);
+	let mut development_preset = runtime_caller
+		.get_named_preset(Some(&sp_genesis_builder::LOCAL_TESTNET_RUNTIME_PRESET.to_string()))
+		.expect("development preset is available on test runtime; qed");
+
+	sc_chain_spec::json_merge(&mut development_preset, patch_json.into());
+
 	ChainSpec::builder(
 		code,
 		Extensions { para_id: id.unwrap_or(cumulus_test_runtime::PARACHAIN_ID.into()).into() },
 	)
 	.with_name("Local Testnet")
-	.with_id("local_testnet")
+	.with_id(sp_genesis_builder::LOCAL_TESTNET_RUNTIME_PRESET)
 	.with_chain_type(ChainType::Local)
-	.with_genesis_config_patch(testnet_genesis_with_default_endowed(
-		extra_endowed_accounts.clone(),
-		id,
-	))
+	.with_genesis_config(development_preset)
```

Comment:
isn't `development_preset` still a patch?

Reply:
it now works because preset is actually full struct json representation. It may change if we go with macro (#5813).